### PR TITLE
add /cloud/join for integration with BlueOS Cloud

### PR DESCRIPTION
**Pull request by Williangalvani**
still WIP, we intend to keep the page hidden while we work on it.

### CODE REVIEW

File: core/frontend/src/views/CloudView.vue [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+export default Vue.extend({
+  name: 'CloudView',
+  data() {
+    return {
+      machineid: undefined,
+      cloud_token: '',
+    }
+  },
+  async mounted() {
+    const result = await commander.command("python -c 'import machineid; print(machineid.hashed_id())'")
```
Comment (patrickelectric) [inline core/frontend/src/views/CloudView.vue:47]: Isn't it better to have helper to provide such information on the frontend over doing python over ts ?

Comment (Williangalvani) [reply core/frontend/src/views/CloudView.vue:47]: it definitely is. actually I was thinking it could be done by linux2rest, what do you think?

Comment (joaoantoniocardoso) [reply core/frontend/src/views/CloudView.vue:47]: > This ID uniquely identifies the host. It should be considered
       "confidential", and must not be exposed in untrusted
       environments, in particular on the network. If a stable unique
       identifier that is tied to the machine is needed for some
       application, the machine ID or any part of it must not be used
       directly. Instead the machine ID should be hashed with a
       cryptographic, keyed hash function, using a fixed,
       application-specific key. That way the ID will be properly
       unique, and derived in a constant way from the machine ID but
       there will be no way to retrieve the original machine ID from the
       application-specific one. The
       [sd_id128_get_machine_app_specific(3)](https://man7.org/linux/man-pages/man3/sd_id128_get_machine_app_specific.3.html) API provides an
       implementation of such an algorithm.


https://man7.org/linux/man-pages/man5/machine-id.5.html

So I guess we should pass an application key to it, like `machineid.hashed_id({blueos})`

To avoid using commander, we could bind the /etc/machine-id as read-only to the BlueOs container.

Comment (Williangalvani) [reply core/frontend/src/views/CloudView.vue:47]: except we dont want to use machine-id, right? since it does a prettye bad job at "uniquely identifying the host"...
I vote that at first we just hash the cpu serial. I have no strong feelings about that being confidential.

For it do actually be confidention, we'd have to implement the "join" logic mostly in the backend, though

Comment (patrickelectric) [reply core/frontend/src/views/CloudView.vue:47]: I agree with pascal about the security concerns.
For now, what I can point is:

For a rasp we can use the serial number, and for everything else we could have custom support or use directly https://docs.rs/machineid-rs/1.2.4/machineid_r, that should support everything necessary for hardware identification generation and add it in linux2rest as pointed by @Williangalvani for easy access API. For more information about crates capabilities, check https://docs.rs/machineid-rs/1.2.4/machineid_rs/enum.HWIDComponent.html

But, that may not fix the issue, since we are going to have only a local source that may not work reliable.

I vote for us to have an automatic request by blueos to blueos.cloud requesting an unique identifier based on the HARDWARE_IDENTIFIER.

```
HARDWARE_IDENTIFIER = `${HARDWARE_IDENTIFIER}-${LINUX_MACHINE_ID}`
BLUEOS_CLOUD_IDENTIFIER = request_blueos_cloud_identifier(HARDWARE_IDENTIFIER)
BLUEOS_IDENTIFIER = `${BLUEOS_CLOUD_IDENTIFIER}-${HARDWARE_IDENTIFIER}`
```

With that we ensure a global unique identifier.

Comment (patrickelectric) [reply core/frontend/src/views/CloudView.vue:47]: `HARDWARE_IDENTIFIER` is generated during execution time, so with that we're going to check if the two computers share the same CLOUD unique identifier with different hardware identification. That'll happen on the case where the sdcard is cloned. 

How `HARDWARE_IDENTIFIER` will be created depend on us. Right now, I would use the pi serial number + `machineid-rs (SystemID, CPUCores, MacAddress, CPUID)`, where the `machineid-rs` would be the standard way of doing it if there are no custom support for the identified system.

Comment (joaoantoniocardoso) [reply core/frontend/src/views/CloudView.vue:47]: > `HARDWARE_IDENTIFIER` is generated during execution time, so with that we're going to check if the two computers share the same CLOUD unique identifier with different hardware identification. That'll happen on the case where the sdcard is cloned.
> 
> How `HARDWARE_IDENTIFIER` will be created depend on us. Right now, I would use the pi serial number + `machineid-rs (SystemID, CPUCores, MacAddress, CPUID)`, where the `machineid-rs` would be the standard way of doing it if there are no custom support for the identified system.

Okay, so you want to do only (B)?

Comment (patrickelectric) [reply core/frontend/src/views/CloudView.vue:47]: > > `HARDWARE_IDENTIFIER` is generated during execution time, so with that we're going to check if the two computers share the same CLOUD unique identifier with different hardware identification. That'll happen on the case where the sdcard is cloned.
> > How `HARDWARE_IDENTIFIER` will be created depend on us. Right now, I would use the pi serial number + `machineid-rs (SystemID, CPUCores, MacAddress, CPUID)`, where the `machineid-rs` would be the standard way of doing it if there are no custom support for the identified system.
> 
> Okay, so you want to do only (B)?

It's `C`, where we keep track of both, we ask `BLUEOS_CLOUD_IDENTIFIER` that is always unique. With this, we are going to be sure that is always a single unique identifier per vehicle.

# SD Card cloned

The `HARDWARE_IDENTIFIER` will not be the same when comparing to `BLUEOS_CLOUD_IDENTIFIER` on the cloud database, we are going to detect this case.

# New SD Card, old registered hardware

The `HARDWARE_IDENTIFIER` will match but `BLUEOS_CLOUD_IDENTIFIER` not once connects to `blueos.cloud`. We are going to detect this case as well.

**Review by voorloopnul — APPROVED** [review]

File: core/frontend/src/store/commander.ts [hunk 2]
```diff
@@ -38,6 +38,28 @@ class CommanderStore {
       })
   }
 
+  async command(command: string): Promise<undefined | ReturnStruct> {
```
Comment (patrickelectric) [inline core/frontend/src/store/commander.ts:41]: Can we break this in another commit ?

Comment (patrickelectric) [inline core/frontend/src/store/commander.ts:41]: Can we break this in another commit ?

File: core/frontend/src/store/commander.ts [hunk 3]
```diff
@@ -38,6 +38,28 @@ class CommanderStore {
...
+        i_know_what_i_am_doing: true,
+      },
+    })
+      .then((response) => response.data)
+      .catch((error) => {
+        if (error === backend_offline_error) {
+          return undefined
+        }
+        const message = `Could not send command to docker: ${error.response?.data ?? error.message}.`
+        console.log(message)
```
Comment (patrickelectric) [inline core/frontend/src/store/commander.ts:57]: please remove development messages

File: core/frontend/src/store/commander.ts [hunk 4]
```diff
@@ -38,6 +38,28 @@ class CommanderStore {
...
+        notifier.pushError('COMMANDER_COMMAND_HOST_FAIL', message, true)
```
Comment (patrickelectric) [inline core/frontend/src/store/commander.ts:58]: This is not command_host but command_blouos

File: core/frontend/src/views/CloudView.vue [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          },
+        },
+      ).catch((error) => {
+        if (error.response.status === 401) {
+          notifier.pushError(
+            'CLOUD_REGISTRATION_ERROR',
+            'Invalid token. Failed to link your vehicle to BlueOS Cloud ',
+            true,
+          )
+        } else if (error.response.status === 400) {
```
Comment (patrickelectric) [inline core/frontend/src/views/CloudView.vue:65]: else is not necessary, an early return can be performed.

Comment (patrickelectric) [inline core/frontend/src/views/CloudView.vue:65]: StatusDode.UNAUTHORIZED

File: core/frontend/src/views/CloudView.vue [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          notifier.pushSuccess(
+            'CLOUD_REGISTRATION_SUCCESS',
+            'Token is valid, your vehicle is now linked to BlueOS Cloud',
+            true,
+          )
+          // TODO: save this securely on a backend or file instead of bag
+          bag.setData('vehicle.cloud', { token: this.cloud_token })
```
Comment (patrickelectric) [inline core/frontend/src/views/CloudView.vue:72]: How this information is going to be used ? Does this PR work ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/frontend/src/views/CloudView.vue (hunk 6 above)
Comment (Williangalvani) [reply core/frontend/src/views/CloudView.vue:73]: For now it just validates the token with the backend and stores it locally. it will only be consumed in subsequent PRs

**Comment by Williangalvani** [comment]
I think we decided to have this be in an extension together with major tom, is that right?

**Comment by patrickelectric** [comment]
> I think we decided to have this be in an extension together with major tom, is that right?

yes

**Comment by voorloopnul** [comment]
I'm _hoping_ to not have a web service running in the major_tom extension, just a daemon, so if we can have the token/auth process on blueos itself, I'll not complain :D 

Also, I think we discussed before that the cloud joining would eventually happen in the wizard, not sure how easy it is to mix things being hosted as an extension with the overall UI of BlueOS.

**Comment by patrickelectric** [comment]
> I'm _hoping_ to not have a web service running in the major_tom extension, just a daemon, so if we can have the token/auth process on blueos itself, I'll not complain :D
> 
> Also, I think we discussed before that the cloud joining would soon happen in the wizard, not sure how easy it is to mix things being hosted as an extension with the overall UI of BlueOS.
> 
> @Williangalvani @patrickelectric

Could we have only the joining / token processes in BlueOS and share that somehow with extensions ?

**Comment by voorloopnul** [comment]
Yes, this PR does precisely that. It validates and saves the token in the bag.vehicle.cloud.
Also, this PR creates a blueOS URL "/cloud/join/" but doesn't advertise it anywhere; it is hidden, so only selected users can know and use it in the early stages, but later should be integrated in the wizard.
